walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,79 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.HashSet;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Set<String> attendees = new HashSet<String>(request.getAttendees());
+    List<TimeRange> initalRanges = new ArrayList<TimeRange>();
+    List<TimeRange> results = new ArrayList<TimeRange>();
+    final int minMeetingTime = 15;
+    
+    //Creates a set of candidate blocks of time, excludes blocks not viable for new meeting,
+    //then consolidates remaining blocks into ranges 
+    for(int start = 0; start <= TimeRange.END_OF_DAY; start += minMeetingTime) {
+        if(start + minMeetingTime > TimeRange.END_OF_DAY) {
+            initalRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, /*inclusivie*/ false));
+        } else {
+            initalRanges.add(TimeRange.fromStartDuration(start, minMeetingTime));
+        }
+    }
+
+    for(Event currEvent : events) {
+        Set<String> currAttendees = currEvent.getAttendees();
+        for(String attendee : currAttendees) {
+            if(attendees.contains(attendee)) {
+                blockTimeRange(initalRanges, currEvent);
+                break;
+            }
+        }
+    }
+
+    TimeRange tempRange = new TimeRange();
+    int duration = (int) request.getDuration();
+
+// Group together the leftover time ranges for results
+    for(TimeRange range : initalRanges) {
+
+        //if range isn't valid then add tempRange to results then reset it
+        if(!range.getValidRange()) {
+            maybeAddToResults(tempRange, results, duration);
+            tempRange.setStart(0);
+            tempRange.setDuration(0);
```

Comment:
I'm not sure I follow why we'd add to results if the range is not valid? Could we instead ensure only valid data makes it into initialRanges in the first place?

It looks like the change overloads the meaning of "duration" to carry along the concept of: does this range represent a free or busy period. As an alternative, what if we were more explicit that a TimeRange tracked both a period of time AND whether it was free/busy?

Reply:
If there was a tempRange with a positive duration and we encounter a range that can't be scheduled then we're adding that tempRange to results with an end point that is right before the beginning of the range that can't be scheduled. 

I had a variable tracking if it was able to be scheduled but changed it per tay's suggestion https://github.com/Sarah-Addo/google-step/pull/14#discussion_r441558703